### ISSUE TITLE
Invalid SQL Server update statement

### ISSUE THREAD

**Issue by esetnik** [simplesamlphp/simplesamlphp#2224]
# **Help us to keep this issue-tracker clean!  For questions or support, please refer to our [mailing lists](https://simplesamlphp.org/lists)**

**Specifics of your environment**
1. Are you acting as SP/IdP/proxy?
SP
3. SimpleSAMLphp: What version are you using?
v2.3.0
5. PHP: What version are you using?
v8.3.11
7. Platform: unix or Windows?
unix
9. Webserver: Apache/Nginx/ISS?
nginx

**Describe the bug**
The syntax for updating the index for sqlsrv driver is incorrect. MS SQL Server does not have `IF EXISTS` syntax.
https://github.com/simplesamlphp/simplesamlphp/blob/master/src/SimpleSAML/Store/SQLStore.php#L135-L139

```

2024-08-30T14:58:20.523Z | Aug 30 10:58:20 simplesamlphp WARNING [ff9946dd77] Database error: array (
-- | --
  | 2024-08-30T14:58:20.523Z | 0 => '42000',
  | 2024-08-30T14:58:20.523Z | 1 => 156,
  | 2024-08-30T14:58:20.523Z | 2 => '[Microsoft][ODBC Driver 18 for SQL Server][SQL Server]Incorrect syntax near the keyword \'IF\'.',
  | 2024-08-30T14:58:20.523Z | )
```

**To Reproduce**
1. Use config like:
```php
    'store.type' => 'sql',

    /*
     * The DSN the sql datastore should connect to.
     *
     * See http://www.php.net/manual/en/pdo.drivers.php for the various
     * syntaxes.
     */
    /* 'store.sql.dsn'                 => 'sqlite:/path/to/sqlitedatabase.sq3', */
    'store.sql.dsn' => 'sqlsrv:Server='.getenv('SQL_HOST').';Database='.getenv('SQL_DATABASE').';Encrypt=true;TrustServerCertificate=true',

    /*
     * The username and password to use when connecting to the database.
     */
    'store.sql.username' => getenv('SQL_USER'),
    'store.sql.password' => getenv('SQL_PASSWORD')
```

**Expected behavior**
sql server driver should only use valid syntax

**Comment by monkeyiq** [comment]
I am not familiar with SQL Server. So grains of salt etc.

 It seems that if we move from using the "ALTER TABLE" to the syntax starting with "DROP INDEX" then it does support the "if exists".

If you have a backup or a copy in a test setup then maybe the following statement instead of the first `ALTER TABLE` will do the trick? This is assuming that there was only one constraint in the old version and the `DROP index IF EXISTS` will then succeed. It would be interesting to execute the subselect manually to see if the resulting index name makes sense first.

```
'DROP INDEX IF EXISTS '
   . '(SELECT CONSTRAINT_NAME '
   . '   FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS '
   . '   WHERE TABLE_NAME=' . $this->prefix . '_tableVersion)'
   . ' ON ' . $this->prefix . '_tableVersion',
```

**Comment by esetnik** [comment]
I don't think you are allowed to use a subquery in the `DROP INDEX IF EXISTS` syntax. However, you should be able to accomplish this with the following:

```php
case 'sqlsrv':
    /**
     * Drop old index and add primary key.
     * NOTE: Retrieve the constraint name dynamically and drop it before adding the primary key.
     */
    $update = [
        // Use dynamic SQL to drop the existing constraint (primary key or index)
        'DECLARE @constraintName NVARCHAR(128); ' .
        'SELECT @constraintName = CONSTRAINT_NAME FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS ' .
        'WHERE TABLE_NAME = \'' . $this->prefix . '_tableVersion\';' .
        'IF @constraintName IS NOT NULL ' .
        'BEGIN ' .
            'EXEC(\'ALTER TABLE ' . $this->prefix . '_tableVersion DROP CONSTRAINT \' + @constraintName); ' .
        'END;',
        
        // Add the new primary key constraint
        'ALTER TABLE ' . $this->prefix . '_tableVersion ADD CONSTRAINT PK_' . $this->prefix . '_tableVersion PRIMARY KEY CLUSTERED (_name);',
    ];
    break;
```

**Comment by monkeyiq** [comment]
This looks good. I will try to get an SQL Server to test against

**Comment by monkeyiq** [comment]
I have made a basic table with
```
CREATE TABLE foo_tableVersion (_name VARCHAR(30) NOT NULL UNIQUE, _version INTEGER NOT NULL);
GO
```
I have seen the error with SQL from the existing code.

I extracted the SQL from this issue to get 
```
DECLARE @constraintName NVARCHAR(128); 
SELECT @constraintName = CONSTRAINT_NAME FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS 
WHERE TABLE_NAME = 'foo_tableVersion' AND CONSTRAINT_TYPE = 'UNIQUE';

IF @constraintName IS NOT NULL
BEGIN 
  EXEC('ALTER TABLE foo_tableVersion DROP CONSTRAINT ' + @constraintName );
END;
ALTER TABLE foo_tableVersion ADD CONSTRAINT PK_foo_tableVersion PRIMARY KEY CLUSTERED (_name);
```

I also had a look at the constraint and pk before and after running the updated SQL above.

```
SELECT CONSTRAINT_NAME FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS 
WHERE TABLE_NAME = 'foo_tableVersion' AND CONSTRAINT_TYPE = 'UNIQUE';

exec sp_pkeys 'foo_tableVersion'
```

**Comment by monkeyiq** [comment]
My double checking thoughts were about the `EXEC` call. On the other hand... all the other schema SQL clauses in `initTableVersionTable()` are inserting `$this->prefix` into the query string directly. 

It may then come down to perhaps doing a regex check on `$this->prefix` in `initTableVersionTable()` to make sure that nothing untoward has somehow got inserted into that variable. It is coming from `store.sql.prefix` in the configuration file. Maybe we could have config::getOptionalString also accept a regex to match with to restrict the valid value range.

The point here is that if somebody can insert an injection of some kind in there they could before and could have used EXEC in that injection if desired. But to do that they would have to either override the local member variable `$prefix` or have write access to the config.php. If they can do the later then the sys admin probably has bigger problems to investigate.

**Comment by tvdijen** [comment]
You could add an extra protection using `Assert::regex()`. Even in a non-attacker scenario, we should make sure no illegal characters are being using here.

**Comment by monkeyiq** [comment]
I have put the update into https://github.com/simplesamlphp/simplesamlphp/pull/2238

I think adding the assert could be done in a subsequent update. Though I also have thoughts of maybe updating the config system instead of the call points for that. I'll open a new issue for it.

**Comment by esetnik** [comment]
The best way to generically prevent against sql injection attacks is never to craft sql queries using any form of string concatenation where variables are supplied as part of the string. Instead, use a prepared statement with bindings, e.g. how it's being done in `->query()` https://github.com/simplesamlphp/simplesamlphp/blob/master/src/SimpleSAML/Database.php#L213-L233

**Comment by monkeyiq** [comment]
I think the `assert` on the `$prefix` and possible move to prepared statements are nice ideas but also shouldn't hold this update back. I have merged it in 2.3 and cherry picked it forward through master.
